cocos/loop/Collections.h

```diff
@@ -0,0 +1,139 @@
+/****************************************************************************
+Copyright (c) 2008-2010 Ricardo Quesada
+Copyright (c) 2009      Valentin Milea
+Copyright (c) 2010-2012 cocos2d-x.org
+Copyright (c) 2011      Zynga Inc.
+Copyright (c) 2013-2016 Chukong Technologies Inc.
+Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.
+
+http://www.cocos2d-x.org
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+****************************************************************************/
+
+
+#pragma once
+
+#include <list>
+#include <vector>
+#include <unordered_map>
+#include <functional>
+
+#include <thread>
+#include <mutex>
+
+#include <cassert>
+
+#ifdef _TMP_CC_LOOP_TS_LOCK
+#error "_TMP_CC_LOOP_TS_LOCK already defined!"
+#endif // _TMP_CC_LOOP_TS_LOCK
+
+
+#define _TMP_CC_LOOP_TS_LOCK std::lock_guard<std::recursive_mutex> guard(_mtx)
+
+namespace cocos2d
+{
+    namespace loop
+    {
+
+        template<typename T>
+        class ThreadSafeQueue {
+        public:
+            void pushBack(T&& ele) { _TMP_CC_LOOP_TS_LOCK; _data.push_back(ele); }
+            void pushBack(T& ele) { _TMP_CC_LOOP_TS_LOCK; _data.push_back(ele); }
+            void pushBack(const T& ele) { _TMP_CC_LOOP_TS_LOCK; _data.push_back(ele); }
+
+            void pushFront(T &ele) { _TMP_CC_LOOP_TS_LOCK; _data.push_front(ele); }
+            void pushFront(const T &ele) { _TMP_CC_LOOP_TS_LOCK; _data.push_front(ele); }
+            void pushFront(T &&ele) { _TMP_CC_LOOP_TS_LOCK; _data.push_front(ele); }
+
+            void popBack() { _TMP_CC_LOOP_TS_LOCK; _data.pop_back(); }
+            T popFront() { _TMP_CC_LOOP_TS_LOCK; assert(_data.size() > 0); T x = _data.front(); _data.pop_front(); return x; }
+
+            T &front() { _TMP_CC_LOOP_TS_LOCK; return _data.front(); }
+            T &back() { _TMP_CC_LOOP_TS_LOCK; return _data.back(); }
```

Comment:
Methods with T& (and others returning &) are not thread safe.
```
thread1:
auto& myValue= ThreadSafeQueue<Value>.back()
thread2:
ThreadSafeQueue<Value>.popBack();
```
Race condition: myValue point to invalid object

Reply:
good catch

Reply:
`ThreadSafeQueue` is designed to protect inner list only, not every element inside.